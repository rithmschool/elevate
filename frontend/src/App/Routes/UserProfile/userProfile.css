.user-salary-form-div {
  margin-bottom: "8em"
}