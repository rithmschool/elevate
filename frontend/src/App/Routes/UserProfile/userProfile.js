import React from "react";
import { Row, Col } from "reactstrap";

import { UserContext } from "../../../userContext";
import ElevateApi from "../../../elevateApi";
import UserSalaryInfoForm from "./UserSalaryInfoForm/userSalaryInfoForm";
import Spinner from "../../../Spinner/spinner";
import Alert from "../Alert/alert";
import UserBasicInfoForm from "./UserBasicInfoForm/userBasicInfoForm";

const MESSAGE_SHOW_PERIOD_IN_MSEC = 3000;

class UserProfile extends React.Component {
  static contextType = UserContext;

  static defaultProps = {
    match: { params: {} }
  };

  constructor(props) {
    super(props);

    this.state = {
      lastestSalary: null,
      isLoading: true,
      errors: [],
      saveConfirmed: false,
      hasSalaryRecord: true
    };

    this.handleUserUpdate = this.handleUserUpdate.bind(this);
    this.handleSalaryUpdate = this.handleSalaryUpdate.bind(this);
  }

  /**
   check for user salary records in salaries table
   set default to 0 if there are no salary records
  **/
  async componentDidMount() {
    const userId = this.context.userId || 0;
    try {
      let lastestSalary = await ElevateApi.getLatestSalary(userId);

      this.setState({ lastestSalary, isLoading: false });
    } catch (err) {
      const lastestSalary = {
        user_id: this.context.userId,
        salary: 0,
        bonus: 0,
        equity: 0
      };

      this.setState({
        lastestSalary,
        isLoading: false,
        hasSalaryRecord: false
      });
    }
  }

  async handleSalaryUpdate(salary) {
    if (!this.state.hasSalaryRecord) {
      try {
        await ElevateApi.postSalary(salary);
        this.setState(
          {
            errors: [],
            saveConfirmed: true,
            isEdit: false
          },
          () => {
            setTimeout(
              () => this.setState({ saveConfirmed: false }),
              MESSAGE_SHOW_PERIOD_IN_MSEC
            );
          }
        ); // remove status message after setTimeout
      } catch (errors) {
        this.setState({ errors });
      }
    } else {
      try {
        await ElevateApi.updateSalary(this.context.userId, salary);
        this.setState(
          {
            errors: [],
            saveConfirmed: true,
            isEdit: false
          },
          () => {
            setTimeout(
              () => this.setState({ saveConfirmed: false }),
              MESSAGE_SHOW_PERIOD_IN_MSEC
            );
          }
        ); // remove status message after setTimeout
      } catch (errors) {
        this.setState({ errors });
      }
    }
  }

  async handleUserUpdate(updateUser, userId) {
    try {
      await ElevateApi.updateUser(userId, updateUser);
      this.setState(
        {
          saveConfirmed: true
        },
        () => {
          setTimeout(
            () => this.setState({ saveConfirmed: false }),
            MESSAGE_SHOW_PERIOD_IN_MSEC
          );
        }
      ); // remove status message after setTimeout
    } catch (errors) {
      this.setState({ errors });
    }
  }

  render() {
    const currentUser = this.context;
    const lastestSalary = this.state.lastestSalary;
    if (this.state.isLoading) return <Spinner />;

    return (
      <div className="container">
        {this.state.errors.length > 0 && (
          <Alert type="danger" messages={this.state.errors} />
        )}

        {this.state.saveConfirmed && (
          <Alert type="success" messages={["Updated successfully."]} />
        )}
        <Row>
          <Col md={6}>
            <div>
              <UserBasicInfoForm
                handleUserUpdate={this.handleUserUpdate}
                handleSalaryUpdate={this.handleSalaryUpdate}
                {...currentUser}
              />
            </div>
          </Col>
        </Row>

        <Row>
          <Col md={6}>
            <div style={{ marginBottom: "8em" }}>
              <UserSalaryInfoForm
                handleSalaryUpdate={this.handleSalaryUpdate}
                {...lastestSalary}
              />
            </div>
          </Col>
        </Row>
      </div>
    );
  }
}

export default UserProfile;
