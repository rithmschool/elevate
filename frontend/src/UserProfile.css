
.sidebar a{
  font-weight: bold;
  font-size: 14px;
  color: #394557;
  display: inline-block;
  position: relative;
  opacity: 0.75;

} 



.sidebar {
  min-width: 250px;
  max-width: 250px;
  height: calc(100vh - 130px);
  overflow-y: scroll;
  margin-left: -350px;
  transition: all 0.5s;
  z-index: 1;
  position: absolute;
}

.sidebar.is-open {
  margin-left: 0;
  transition: .5s;
}
/* li:hover {
  background: #7e8b9f;
  }

.sidebar > a:hover{
  color: aliceblue;
} */

@media only screen and (max-width: 500px) {

  .content.is-open {
    margin-left: 100%;
  }

  .sidebar.is-open {
    min-width: 100%;
    max-width: 100%;
    margin-left: 0;
    transition: all 0.5s, height 0s;
  }

}