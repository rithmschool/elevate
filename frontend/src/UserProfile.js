import React from "react";
import {UserContext} from "./UserContext"
import UserBasicInfoForm from './UserBasicInfoForm';
import UserSalaryInfoForm from './UserSalaryInfoForm'
import ElevateApi from './ElevateApi';
import { Row, Col } from 'reactstrap';
import Spinner from './Spinner';
import Alert from "./Alert";


const MESSAGE_SHOW_PERIOD_IN_MSEC = 3000;

class UserProfile extends React.Component {
  static contextType = UserContext;
  static defaultProps = {
    match: { params: {} },
  };
  constructor(props) {
    super(props);
    this.state = {
      lastestSalary: null,
      isLoading: true,
      errors: [],
      saveConfirmed: false,
      hasSalaryRecord: true,
    }
    this.handleUserUpdate = this.handleUserUpdate.bind(this);
    this.handleSalaryUpdate = this.handleSalaryUpdate.bind(this);
    

  }


  /**
   * check if the user has any salary records in salaries table
   * if No set hasSalaryRecord to false and create latestSalary object with default values of 0
   *  */
  async componentDidMount(){
    const userId = this.context.userId;
    try{
      let lastestSalary = await ElevateApi.getLatestSalary(userId);
      this.setState({lastestSalary, isLoading: false});
    } catch(err){
      // if current user has not any salaries record create an empty salary entry to populate the display form
        const lastestSalary = {
          user_id: this.context.userId,
          salary: 0,
          bonus: 0,
          equity: 0
        }
        this.setState({lastestSalary, isLoading: false, hasSalaryRecord: false});
      }
  }

  async handleSalaryUpdate(salary){
    if(!this.state.hasSalaryRecord){
      try{
        let res = await ElevateApi.postSalary(salary);
        this.setState(
          {
            errors: [],
            saveConfirmed: true,
            isEdit: false,
          }, () => {
            // after a short period, remove status message
            setTimeout(
              () => this.setState({ saveConfirmed: false }),
              MESSAGE_SHOW_PERIOD_IN_MSEC);
          });
      } catch (errors) {
          this.setState({ errors });
        }
    }
    else {
        try{
          await ElevateApi.updateSalary(this.context.userId, salary);
          this.setState(
            {
              errors: [],
              saveConfirmed: true,
              isEdit: false,
            }, () => {
              // after a short period, remove status message
              setTimeout(
                () => this.setState({ saveConfirmed: false }),
                MESSAGE_SHOW_PERIOD_IN_MSEC);
            });
        } catch (errors) {
            this.setState({ errors });
          }
        }
}
// update user infos
  async handleUserUpdate(updateUser,userId){
    try{
      await ElevateApi.updateUser(userId, updateUser);

      this.setState(
        {
          saveConfirmed: true,
        }, () => {
          // after a short period, remove status message
          setTimeout(
            () => this.setState({ saveConfirmed: false }),
            MESSAGE_SHOW_PERIOD_IN_MSEC);
        });
    } catch (errors) {
        this.setState({ errors });
      }
  }
     
   
  render(){
    const currentUser = this.context;
    const lastestSalary = this.state.lastestSalary;
    if(this.state.isLoading)return <Spinner/>;

    return (
      <div className="container">
        {this.state.errors.length > 0 &&
              <Alert type="danger" messages={this.state.errors} />}

            {this.state.saveConfirmed &&
              <Alert type="success"
                    messages={["Updated successfully."]} />}
        <Row>
         <Col md={6}>
            <h3 className="text-info">User information</h3>
          </Col>
        </Row>
        <br></br>
        <Row>
          <Col md={6}>
          <div >
            <UserBasicInfoForm
              handleUserUpdate={this.handleUserUpdate}
              handleSalaryUpdate={this.handleSalaryUpdate}
              {...currentUser}
              />
          </div>
          </Col>
        </Row>
        <br></br>
        <Row>
          <Col md={6}>
          <div >

            <UserSalaryInfoForm
              handleSalaryUpdate={this.handleSalaryUpdate}
              {...lastestSalary}
              />
          </div>
          </Col>
          </Row>
      </div>  
    )
  } 
}
export default UserProfile;
