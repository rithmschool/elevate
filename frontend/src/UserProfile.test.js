
import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import UserProfile from "./UserProfile";

it("renders without crashing", function () {
  shallow(<UserProfile />);
});

it("matches snapshot", function () {
  let wrapper = shallow(<UserProfile />);
  let serialized = toJson(wrapper);
  expect(serialized).toMatchSnapshot();
});
