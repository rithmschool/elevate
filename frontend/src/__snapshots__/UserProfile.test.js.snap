// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="EditPUserForm container border rounded"
  style={
    Object {
      "backgroundColor": "#F4F6F8",
    }
  }
>
  <Row
    tag="div"
  >
    <Col
      md={6}
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <h3
        className="text-info"
      >
        User information
      </h3>
    </Col>
    <Col
      align="right"
      md={6}
      sm={
        Object {
          "offset": 0,
          "size": 8,
        }
      }
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Button
        color="info"
        onClick={[Function]}
        size="sm"
        style={
          Object {
            "marginTop": "6px",
          }
        }
        tag="button"
      >
        Edit details
      </Button>
    </Col>
  </Row>
  <hr />
  <Form
    onSubmit={[Function]}
    tag="form"
  >
    <Row
      form={true}
      tag="div"
    >
      <Col
        md={6}
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <Label
          className="form-group has-float-label"
          tag="label"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
        >
          <Input
            disabled={true}
            id="EditUser-first_name"
            name="first_name"
            onChange={[Function]}
            type="text"
          />
          <span>
            First name
          </span>
        </Label>
      </Col>
      <Col
        md={6}
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <Label
          className="form-group has-float-label"
          tag="label"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
        >
          <Input
            disabled={true}
            id="EditUser-last_name"
            name="last_name"
            onChange={[Function]}
            type="text"
          />
          <span>
            Last name
          </span>
        </Label>
      </Col>
    </Row>
    <Row
      form={true}
      tag="div"
    >
      <Col
        md={6}
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <Label
          className="form-group has-float-label"
          tag="label"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
        >
          <Input
            disabled={true}
            id="EditUser-email"
            name="email"
            onChange={[Function]}
            type="email"
          />
          <span>
            Email
          </span>
        </Label>
      </Col>
      <Col
        md={6}
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <Label
          className="form-group has-float-label"
          tag="label"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
        >
          <Input
            disabled={true}
            id="EditUser-current_company"
            name="current_company"
            onChange={[Function]}
            type="text"
          />
          <span>
            Current company
          </span>
        </Label>
      </Col>
    </Row>
    <Row
      form={true}
      tag="div"
    >
      <Col
        md={6}
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <Label
          className="form-group has-float-label"
          tag="label"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
        >
          <Input
            disabled={true}
            id="EditUser-salary"
            name="salary"
            onChange={[Function]}
            type="number"
          />
          <span>
            salary
          </span>
        </Label>
      </Col>
      <Col
        md={6}
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <Label
          className="form-group has-float-label"
          tag="label"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
        >
          <Input
            disabled={true}
            id="EditUser-hire_date"
            name="hire_date"
            onChange={[Function]}
            type="date"
            value="2019-08-29"
          />
          <span>
            Hiring date
          </span>
        </Label>
      </Col>
    </Row>
    <hr />
    <Col
      align="left"
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <p
        className="text-primary"
      />
      <Button
        color="info"
        disabled={true}
        size="sm"
        tag="button"
      >
        Save details
      </Button>
    </Col>
    <br />
  </Form>
</div>
`;
